### PR TITLE
Include `experimental.jet` into docs

### PR DESCRIPTION
**Pull request by pnkraemer**
This PR introduces a module-level docstring for `jax.experimental.jet.py` and a function-level docstring for `jet.jet()`.

See the issue #3730 and the discussion #9536 for context.

The content of the docstrings is tightly coupled to how I use the API and how I understand the maths based on https://openreview.net/pdf?id=SkxEF3FNPH (as well as Griewank/Walther). Please let me know if I should rephrase some statements. I will leave comments in the diff, where I explain the thought process behind specific statements.

### CODE REVIEW

File: jax/experimental/jet.py [hunk 1]
```diff
@@ -12,6 +12,41 @@
...
+
+  More specifically, :func:`jet` computes
+
+  .. math::
+    y_0, (y_1, . . . , y_d) = \texttt{jet} (f, x_0, (x_1, . . . , x_d))
+
+  and can thus be used for high-order
+  automatic differentiation of :math:`f`.
+  Details are explained in
+  `this paper <https://openreview.net/forum?id=SkxEF3FNPH>`__.
```
Comment (pnkraemer) [inline jax/experimental/jet.py:43]: The content of this docstring closely follows the exposition in https://openreview.net/pdf?id=SkxEF3FNPH. It seems to be more maths than other, comparable module docstrings, but I figured it is useful to establish the variables for easier reference of what `jet` does.

File: jax/experimental/jet.py [hunk 2]
```diff
@@ -12,6 +12,41 @@
...
+
+  Note:
+    Help improve :func:`jet` by contributing
+    `outstanding primitive rules <https://github.com/google/jax/issues/2431>`__.
```
Comment (pnkraemer) [inline jax/experimental/jet.py:47]: Other experimental modules had similar comments: `ann.py` lists todos, or `host_callback` states "please give feedback, and expect changes.". If I should remove this comment, though, please let me know.

File: jax/experimental/jet.py [hunk 3]
```diff
@@ -32,6 +67,51 @@
...
+      array, scalar, or standard Python container of arrays or scalars.
+    primals: The primal values at which the Taylor approximation of ``fun`` should be
+      evaluated. Should be either a tuple or a list of arguments,
+      and its length should be equal to the number of positional parameters of
+      ``fun``.
+    series: Higher order Taylor-series-coefficients.
+      Together, `primals` and `series` make up a truncated Taylor polynomial.
+      Should be either a tuple or a list of tuples or lists,
+      and its length dictates the degree of the truncated Taylor polynomial.
+
```
Comment (pnkraemer) [inline jax/experimental/jet.py:84]: I did not document the exceptions that are raised, because I did not see this kind of doc in many other functions. But I am aware that it would go here, please let me know if I should add it.

File: jax/experimental/jet.py [hunk 4]
```diff
@@ -32,6 +67,51 @@
...
+  :func:`jet` returns the Taylor coefficients of :math:`f(x(t)) = \sin(t^3)`:
+
+  >>> print(y0,  np.sin(x0))
+  0.12467473 0.12467473
+
+  >>> print(y1, np.cos(x0) * x1)
+  0.7441479 0.74414825
+
+  >>> print(y2, -np.sin(x0) * x1 +  x2)
+  2.9064622 2.906494
```
Comment (pnkraemer) [inline jax/experimental/jet.py:113]: In order to make, for instance, a `doctest` pass reliably, I would have to round these values. But I didnt see much rounding happening similar examples in other modules, and `pytest -n auto jax/experimental/jet.py` does not check these examples at least.

File: jax/experimental/jet.py (hunk 1 above)
Comment (mattjj) [reply jax/experimental/jet.py:43]: Thanks for pushing on this! 

I don't recommend following the math or notation in that workshop paper. I recommend working from [these notes](https://github.com/google/jax/files/6717197/jet.pdf). In particular, the differentiation notation is different, with no \frac. We should link to that pdf as well (instead of openreview).

**Comment by mattjj** [comment]
This is really awesome, thanks so much for contributing this!

File: jax/experimental/jet.py (hunk 3 above)
Comment (mattjj) [reply jax/experimental/jet.py:84]: Not documenting exceptions sounds good to me.

File: jax/experimental/jet.py (hunk 2 above)
Comment (mattjj) [reply jax/experimental/jet.py:47]: This seems like a good idea!

File: jax/experimental/jet.py (hunk 1 above)
Comment (pnkraemer) [reply jax/experimental/jet.py]: Good to know, thanks! I updated the link as well as the notation to be more in line with the notes you linked.

File: jax/experimental/jet.py [hunk 5]
```diff
@@ -12,6 +12,46 @@
...
+    (h(x), \partial h(x)[v], \partial^2 h(x)[v], ..., \partial^K h(x)[v,...,v]),
+
+  which represents a :math:`K`-th order Taylor approximation
+  of :math:`h` at :math:`x`, :func:`jet` returns a :math:`K`-th order
+  Taylor approximation of :math:`f` at :math:`x`,
+
+  .. math::
+    (f_0, ..., f_K) :=
+    (f(x), \partial f(x)[v], \partial^2 f(x)[v], ..., \partial^K f(x)[v,...,v]).
+
```
Comment (pnkraemer) [inline jax/experimental/jet.py:39]: the PDF you linked in the other comment discusses normalised Taylor coefficients (i.e. f_k / k!), but as far as I am aware, `jet` does not work with normalised coefficients. Right? 

Because if it does, I should update this docstring! :)

Comment (mattjj) [reply jax/experimental/jet.py]: > the PDF you linked in the other comment discusses normalised Taylor coefficients (i.e. f_k / k!), but as far as I am aware, jet does not work with normalised coefficients. Right?

Hah, I forget! Thanks for checking. It would be good to make sure the docstring agrees with the code. If you've looked at the code recently, I trust your evaluation as to which it uses!

**Comment by YouJiacheng** [comment]
I would appreciate it if you could add some multivariate examples? For example, how to compute hessian of a MLP using `jet`? Or it is impossible to compute hessian accelerated by `jet`?🥺

**Comment by pnkraemer** [comment]
> I would appreciate it if you could add some multivariate examples? For example, how to compute hessian of a MLP using `jet`? Or it is impossible to compute hessian accelerated by `jet`?pleading_face

No, I don't think it is impossible at all! :) 

But I am not sure the example in the documentation is the best place for it. I think it should also not be done as part of this pull request, but I will include it if @mattjj wants me to :) 

There seem to be a few more or less ongoing discussions about documentation of `jet`: e.g., #3730, #9536, and #9598 (which you raised -- see also my reply there for some more context). If those discussions keep attracting interest, perhaps someone volunteers to put together more elaborate examples of this type of functionality?

**Comment by mattjj** [comment]
Hey, sorry for going silent on this. I got drawn into unrelated stuff.

**Review by mattjj — APPROVED** [review]
Looks great! Thanks for this contribution. I think it'll help rebuild momentum around `jet`!
